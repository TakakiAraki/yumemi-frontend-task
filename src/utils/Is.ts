import { MEDIA } from "~/constants/style";

export const is = {
  production: process.env.NODE_ENV === "production",
  development: process.env.NODE_ENV === "development",
  test: process.env.NODE_ENV === "test",
  null: <T extends unknown>(value: T | null | undefined): value is null | undefined =>
    value == null,
  notNull: <T extends unknown>(value: T | null | undefined): value is T => value != null,
  client: typeof process === "undefined",
  server: typeof window === "undefined",
  array: Array.isArray,
  isPC: () => window.matchMedia(MEDIA.IS_SP).matches,
  isSP: () => !window.matchMedia(MEDIA.IS_SP).matches,
};
