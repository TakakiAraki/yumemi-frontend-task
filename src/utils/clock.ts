interface ClockState {
  days: number;
  hours: number;
  minutes: number;
  seconds: number;
  milliseconds: number;
}
const initialState: ClockState = {
  days: 0,
  hours: 0,
  minutes: 0,
  seconds: 0,
  milliseconds: 0,
};

export const clock = (state: Partial<ClockState>) => {
  return {
    toSeconds: () => {
      const result = { ...initialState, ...state };
      return (
        result.milliseconds / 1000 +
        result.seconds +
        result.minutes * 60 +
        result.hours * 60 * 60 +
        result.days * 60 * 60 * 24
      );
    },
    toMilliseconds: () => {
      const result = { ...initialState, ...state };
      return (
        result.milliseconds +
        result.seconds * 1000 +
        result.minutes * 1000 * 60 +
        result.hours * 1000 * 60 * 60 +
        result.days * 1000 * 60 * 60 * 24
      );
    },
  };
};
