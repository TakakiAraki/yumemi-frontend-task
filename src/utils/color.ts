import diff from "color-diff";

// ref https://stackoverflow.com/questions/3426404/create-a-hexadecimal-colour-based-on-a-string-with-javascript
interface Color {
  R: number; // 255
  G: number; // 255
  B: number; // 255
}
const parseColor = (str: string): Color => {
  const target = str.replace("#", "");
  switch (target.length) {
    case 8:
      return {
        R: parseInt(target[0] + target[1], 16),
        G: parseInt(target[2] + target[3], 16),
        B: parseInt(target[4] + target[5], 16),
      };
    case 6:
      return {
        R: parseInt(target[0] + target[1], 16),
        G: parseInt(target[2] + target[3], 16),
        B: parseInt(target[4] + target[5], 16),
      };
    case 3:
      return {
        R: parseInt(target[0] + target[0], 16),
        G: parseInt(target[1] + target[1], 16),
        B: parseInt(target[2] + target[2], 16),
      };
    default:
      throw new Error("変換対象のstringが不正なフォーマットです");
  }
};

/**
 * @param props
 * @returns
 */
export const color = (props: Color | string) => {
  const colorObject = typeof props === "string" ? parseColor(props) : props;
  return {
    value: colorObject,

    // 色距離が近いものを取得する
    closest: (target: Color[]) => {
      const result = diff.closest(colorObject, target);
      return color(result);
    },

    // 色距離が遠いものを取得する
    futhest: (target: Color[]) => {
      const result = diff.furthest(colorObject, target);
      return color(result);
    },

    // #FFFFFF #FFFFFFFF 形式で出力する
    toRGBString: () => {
      const fillZero = (str: string) => ("00" + str).slice(-2);
      return (
        "#" +
        fillZero(colorObject.R.toString(16)) +
        fillZero(colorObject.G.toString(16)) +
        fillZero(colorObject.B.toString(16))
      );
    },
  };
};
export type TColor = typeof color;
export const BLACK: Color = color("#000000").value;
export const WHITE: Color = color("#FFFFFF").value;

/**
 * 文字列を色に直すメソッド
 * @param str ハッシュ化させる文字列
 * @returns RGBの値
 */
export const stringToHashColor = (str: string) => {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    hash = str.charCodeAt(i) + ((hash << 5) - hash);
  }
  const c: number[] = [];
  for (let i = 0; i < 4; i++) {
    const value = (hash >> (i * 8)) & 0xdf;
    c.push(value);
  }

  return color({
    R: c[0],
    G: c[1],
    B: c[2],
  });
};
