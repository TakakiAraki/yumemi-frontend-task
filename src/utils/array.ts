import { is } from "./Is";

type FilterCb<Value extends unknown, Result extends unknown> = (
  value: Value,
  index: number,
  array: Value[],
) => false | null | undefined | Result;

export const reduces = {
  filterMap:
    <Value extends unknown, Result extends unknown>(cb: FilterCb<Value, Result>) =>
    (
      previousValue: Result[],
      currentValue: Value,
      currentIndex: number,
      array: Value[],
    ): Result[] => {
      const value = cb(currentValue, currentIndex, array);
      if ((typeof value === "boolean" && !value) || is.null(value)) return previousValue;
      return [...previousValue, value];
    },
};
