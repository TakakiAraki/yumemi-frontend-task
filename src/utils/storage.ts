import { is } from "./Is";

export const createStorage = <T extends any>(key: string) => {
  return {
    // APIである可能性も想定し、Promiseとする
    save: async (data: T): Promise<boolean> => {
      window.localStorage.setItem(key, JSON.stringify(data));
      return true;
    },
    load: async (): Promise<T | undefined> => {
      const result = window.localStorage.getItem(key);
      if (is.null(result)) return;
      return JSON.parse(result);
    },
  };
};
