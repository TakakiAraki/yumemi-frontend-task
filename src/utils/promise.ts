import { clock } from "./clock";
import { is } from "./Is";

/**
 * @param time - 待ち時間 ms
 * @returns
 */
export const waitFor = (time: number) => new Promise((resolve) => setTimeout(resolve, time));

interface PromiseControlOptions {
  thresholdTime?: number;
  maxRequestCount?: number;
}
export const promiseControl = <Callback extends (...args: any[]) => any>(
  cb: Callback,
  options?: PromiseControlOptions,
) => {
  const thresholdTime = options?.thresholdTime ?? clock({ seconds: 1 }).toMilliseconds();
  const maxRequestCount = options?.maxRequestCount ?? 5;
  let counts: number[] = [];

  const result = async (...args: Parameters<Callback>): Promise<ReturnType<Callback>> => {
    const currentTime = Date.now();

    // thresholdTime秒以上経過したものは除外
    counts = counts.filter((val) => currentTime - val <= thresholdTime);

    // wait time を maxRequestCount個数前から算出する
    const targetTime = counts[counts.length - maxRequestCount];
    const waitTime = is.null(targetTime)
      ? 0
      : Math.max(thresholdTime - (currentTime - targetTime), 0);
    counts.push(Date.now() + waitTime);
    await waitFor(waitTime);
    return cb(...args);
  };

  return result;
};
