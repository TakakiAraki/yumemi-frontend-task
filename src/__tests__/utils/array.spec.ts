import { reduce } from "~/utils/array";

describe("reduce test", () => {
  test("filter map", () => {
    const value = [1, 2, 3];
    const initialValue: { val: number }[] = [];

    const result = value.reduce(
      reduce.filterMap((val) => {
        if (val === 1) return false;
        return {
          val,
        };
      }),
      initialValue,
    );
    expect(result).toMatchSnapshot();
  });
});
