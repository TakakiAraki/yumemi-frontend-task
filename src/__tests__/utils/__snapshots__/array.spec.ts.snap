// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reduce test filter map 1`] = `
[
  {
    "val": 2,
  },
  {
    "val": 3,
  },
]
`;
